source/extensions/filters/http/header_mutation/header_mutation.cc

```diff
@@ -28,20 +28,33 @@ PerRouteHeaderMutation::PerRouteHeaderMutation(const PerRouteProtoConfig& config
     : mutations_(config.mutations()) {}
 
 HeaderMutationConfig::HeaderMutationConfig(const ProtoConfig& config)
-    : mutations_(config.mutations()) {}
+    : mutations_(config.mutations()),
+      most_specific_header_mutations_wins_(config.most_specific_header_mutations_wins()) {}
 
 Http::FilterHeadersStatus HeaderMutation::decodeHeaders(Http::RequestHeaderMap& headers, bool) {
   Formatter::HttpFormatterContext ctx{&headers};
   config_->mutations().mutateRequestHeaders(headers, ctx, decoder_callbacks_->streamInfo());
 
-  // Only the most specific route config is used.
-  // TODO(wbpcode): It's possible to traverse all the route configs to merge the header mutations
-  // in the future.
-  route_config_ =
-      Http::Utility::resolveMostSpecificPerFilterConfig<PerRouteHeaderMutation>(decoder_callbacks_);
-
-  if (route_config_ != nullptr) {
-    route_config_->mutations().mutateRequestHeaders(headers, ctx, decoder_callbacks_->streamInfo());
+  // Traverse through all route configs to retrieve all available header mutations.
+  // `getAllPerFilterConfig` returns in ascending order of specificity (i.e., route table
+  // first, then virtual host, then per route).
+  route_configs_ = Http::Utility::getAllPerFilterConfig<PerRouteHeaderMutation>(decoder_callbacks_);
+
+  if (!config_->mostSpecificHeaderMutationsWins()) {
+    // most_specific_wins means that most specific level per filter config is evaluated last. In
+    // other words, header mutations are evaluated in ascending order of specificity (same order as
+    // `getAllPerFilterConfig` above returns).
+    // Thus, here we reverse iterate the vector when `most_specific_wins` is false.
+    for (auto it = route_configs_.rbegin(); it != route_configs_.rend(); ++it) {
+      ASSERT(*it != nullptr);
```

Comment:
Probably don't need this, it's one of the style guide ASSERT anti-patterns (since it would crash next line anyway and is implied by that code).

Reply:
make sense. Done.
I think we should avoid crash on data plane if possible. Actually [getAllPerFilterConfig](https://github.com/envoyproxy/envoy/blob/d5de0df4e2cbc452df9e89864ccc6410ea109197/source/common/http/utility.h#L645-L648)  here would not be nullptr (i.e., would not crash). Assert is redundant.